crates/core/asset/src/asset/denom_metadata.rs

```diff
@@ -317,16 +318,19 @@ impl Metadata {
         if amount == 0u64.into() {
             return self.default_unit();
         }
+        let mut selected_index = 0;
+        let mut selected_exponent = 0;
         for (unit_index, unit) in self.inner.units.iter().enumerate() {
             let unit_amount = Amount::from(10u128.pow(unit.exponent as u32));
-            if amount >= unit_amount {
-                return Unit {
-                    unit_index,
-                    inner: self.inner.clone(),
-                };
+            if unit_amount <= amount && unit.exponent >= selected_exponent {
+                selected_index = unit_index;
+                selected_exponent = unit.exponent;
             }
         }
-        self.base_unit()
+        return Unit {
+            unit_index: selected_index,
+            inner: self.inner.clone(),
+        };
```

Comment:
This bugfix needs a unit test.

It can be triggered by having a Metadata object where the list of units is not sorted. The previous code is incorrect on, for instance, the USDC Metadata currently in the Prax registry. That JSON object could be copied into a unit test, which can create a value like "1.234 USDC", parse it, and check that displaying it round trips (because the correct unit is selected).